### ISSUE TITLE
Can we ignore non-external, non-annotated JS Interop members?

### ISSUE THREAD

**Issue by biggs0125** [dart-lang/sdk#49428]
We have some cases where there are JS interop class members that are not external and not annotated with @JS(<name>) themselves. For example, default constructors of @JS() classes which by definition the user does not specify. We consider the name of these `null` and store them alongside the members which are annotated or external. When the annotation is present on a member but the name is not specified the name gets stored as an empty string. When the member is external but no annotation is provided then the name is defaulted to the Kernel IR member name.

We should consider whether or not it's worth ignoring these non-external, non-annotated members when processing JS Interop classes. There are already some members we ignore: https://github.com/dart-lang/sdk/blob/main/pkg/compiler/lib/src/js_backend/native_data.dart#L197

It's unclear whether or not there's a valid use case for compiling this specific subset of members but if there's not it might make sense to ignore it in that above line and then make the map where we persist those names have a non-null value type:
https://github.com/dart-lang/sdk/blob/main/pkg/compiler/lib/src/js_backend/native_data.dart#L40

**Comment by srujzs** [comment]
I don't believe we have a specification of behavior for synthetic constructors on an interop class. In fact, I'd argue the use of such a constructor should be an error, but I don't know what the current behavior is for these members. If they error on both compilers, I'm fine with ignoring them/doing whatever with them.

`tests/lib/js/external_member_static_test.dart` contains tests for the disallowed members in a `@JS` class. Worth noting are:

- non-external factories: These should be allowed (ignoring the fact that they don't work in DDC). They're useful for creating variants of constructors.
- non-external non-annotated (and annotated) instance fields: This is an unfortunate relic that we've been keeping until we fix all the current usages. These members are equivalent to external instance fields. There's a longstanding bug to handle these usages we have not gotten around to. Since they are explicitly disallowed in static interop, there's a good chance these will just become obsolete.
- non-external non-annotated static members: These should be allowed (and like factories, I believe they don't work in DDC).

We also don't do a great job making sure `@JS` annotations are used only on external members (as they're useless otherwise). Like non-external instance fields, there's no point to these annotations, but we haven't gone through the breaking change process of disallowing them.

I hope this helps. Let me know if you have more questions.

**Comment by biggs0125** [comment]
That all makes sense, thanks! Beyond what we allow and disallow, when we're actually processing these interop members it seems like we can ignore a large subset of those fields. We don't need to track their names to include in the compiled JS.

Specifically the answer to this [TODO](https://github.com/dart-lang/sdk/blob/main/pkg/compiler/lib/src/js_backend/native_data.dart#L193) seems to be that we ignore most of the entities where the name would be null. Can we generalize that membersToIgnore to capture this and add any exclusions if we need them?

**Comment by srujzs** [comment]
> Specifically the answer to this [TODO](https://github.com/dart-lang/sdk/blob/main/pkg/compiler/lib/src/js_backend/native_data.dart#L193) seems to be that we ignore most of the entities where the name would be null.

Yes, that's reasonable. The only exception I can think of is an unnamed external constructor that is non-synthetic. I don't know what the CFE guarantees there but the analyzer has a `null` name for those constructors, so we might want to double-check that.

> Can we generalize that membersToIgnore to capture this and add any exclusions if we need them?

If the goal is to just make the above map non-nullable, then I believe the only possible way a name could be `null` for an interop member is the above constructor.
